### PR TITLE
Fix resource access checks in getResource and getResourceAsStream

### PR DESCRIPTION
**Pull request by pdbain-ibm**
Allow getResource et al. to search the current module if the package in
question is accessible by the caller's module.

Fixes https://github.com/eclipse/openj9/issues/3136

Signed-off-by: Peter Bain <peter_bain@ca.ibm.com>

### CODE REVIEW

File: jcl/src/java.base/share/classes/java/lang/Class.java [hunk 1]
```diff
@@ -1964,25 +1966,66 @@ public URL getResource(String resName) {
...
+ * Look up the resource in the module if the module is named 
+ * and is the same module as the caller or the package is open to the caller.
+ * The default package (i.e. resources at the root of the module) is considered open.
+ * 
+ * @param absoluteResName name of resource, including package
+ * @param thisModule module of the current class
+ * @param callerClass class of method calling getResource() or getResourceAsStream()
+ * @return true if modular lookup should be used.
+ */
+private boolean useModularSearch(String absoluteResName, Module thisModule, Class callerClass) {
```
Comment (DanHeidinga) [inline jcl/src/java.base/share/classes/java/lang/Class.java]: `Class callerClass` -> `Class<?> callerClass`

File: jcl/src/java.base/share/classes/java/lang/Class.java [hunk 2]
```diff
@@ -1964,25 +1966,66 @@ public URL getResource(String resName) {
...
+	boolean visible = false;
+
+	if (thisModule.isNamed()) {
+		final Module callerModule = callerClass.getModule();
+		visible = (thisModule == callerModule);
+		if (!visible) {
+			if (absoluteResName.endsWith(".class")) {
```
Comment (DanHeidinga) [inline jcl/src/java.base/share/classes/java/lang/Class.java]: To match the pattern of code above, use:

```java
visible = absoluteResName.endsWith(".class");
if (!visible) {.....
```

File: jcl/src/java.base/share/classes/java/lang/Class.java [hunk 3]
```diff
@@ -1964,25 +1966,66 @@ public URL getResource(String resName) {
...
+				visible = true;
+			} else {
+				// extract the package name
+				int lastSlash = absoluteResName.lastIndexOf('/');
+				if (-1 == lastSlash) { // no package name
+					visible = true;
+				} else {
+					String result = absoluteResName.substring(0, lastSlash).replace('/', '.');
+					visible = thisModule.isOpen(result, callerModule);
```
Comment (DanHeidinga) [inline jcl/src/java.base/share/classes/java/lang/Class.java]: Does `isOpen` cover all cases or should there also be an `isExported` check?

Comment (pdbain-ibm) [reply jcl/src/java.base/share/classes/java/lang/Class.java]: From the javadoc for `isExported()`:
> A package that is open to the given module is considered exported to that module at run-time and so this method returns true if the package is open to the given module.

so `isExported()` would be redundant.

File: jcl/src/java.base/share/classes/java/lang/Class.java (hunk 2 above)
Comment (pdbain-ibm) [reply jcl/src/java.base/share/classes/java/lang/Class.java]: Good idea. Done.

File: jcl/src/java.base/share/classes/java/lang/Class.java (hunk 1 above)
Comment (pdbain-ibm) [reply jcl/src/java.base/share/classes/java/lang/Class.java]: Right.  Done.
I will push the updates when my personal build completes.

File: jcl/src/java.base/share/classes/java/lang/Class.java [hunk 4]
```diff
@@ -1927,31 +1927,33 @@ String getPackageName() {
...
 		}
-	} else
-/*[ENDIF] Sidecar19-SE */
+	}
+	if (null == result)
+		/*[ENDIF] Sidecar19-SE */
 	{
 		if (loader == ClassLoader.bootstrapClassLoader) {
-			return ClassLoader.getSystemResource(absoluteResName);
+			result =  ClassLoader.getSystemResource(absoluteResName);
```
Comment (DanHeidinga) [inline jcl/src/java.base/share/classes/java/lang/Class.java]: nitpick - there's some odd double spacing here and throughout the new code.  Can you address that before the merge?

Comment (pdbain-ibm) [reply jcl/src/java.base/share/classes/java/lang/Class.java]: Fixed in 5 locations.

**Comment by DanHeidinga** [comment]
Jenkins test sanity plinux jdk8,11

**Comment by DanHeidinga** [comment]
Jenkins test extended xlinux jdk10

**Comment by DanHeidinga** [comment]
Jenkins test sanity plinux jdk8,jdk11

**Comment by pdbain-ibm** [comment]
@DanHeidinga all tests passed.

**Pull request merged**